### PR TITLE
Instant Search: Integrate new Redux store with component tree

### PR DESCRIPTION
**Pull request by jsnmoon**
Follow-up for #18137.

#### Changes proposed in this Pull Request:
* Integrates the new Redux store with the component tree using React-Redux.
* Prunes the query string library to handle only getting and setting the query string. 
* Modify our Webpack configuration to enable aliasing React imports to `preact/compat`.
* Adds `redux` and `react-redux` as dependencies, which increases bundle bloat by 7.6 kB gzipped.
(Before this PR: 40.12 kB gzipped. After this PR: 47.59 kB gzipped.)

#### Does this pull request change what data or activity we track or use?
None.

#### Testing instructions:
* Apply this change to your Jetpack site with Jetpack Search.
* Try testing the following flows:
  1. Spawn the overlay by submitting something into a search input.
  2. Spawn the overlay by clicking on a filter that doesn't exist inside the overlay sidebar.
  3. Spawn the overlay by clicking on a filter that also exists inside the overlay sidebar.
  4. After spawning the overlay, try changing your search query.
  5. After spawning the overlay, try changing your sort type.
  6. After spawning the overlay, try changing your filters.
  7. Try opening a search URL from a new tab (e.g. `?s=home&category=uncategorized`).

For each of the flows, ensure that:
  - The search interface works as expected.
  - The address bar updates when the search query, sort, or filters are updated.

#### Proposed changelog entry for your changes:
* None.

### CODE REVIEW

File: modules/search/instant-search/components/search-app.jsx
Comment (jsnmoon) [inline modules/search/instant-search/components/search-app.jsx:223]: This might no longer be true; will investigate.

File: modules/search/instant-search/components/search-app.jsx
Comment (bluefuton) [inline modules/search/instant-search/components/search-app.jsx:288]: 💯

**Comment by bluefuton** [comment]
As this is such a significant change, feels like it might be worth holding off until after the Jan code freeze to merge it. What do you think @jsnmoon?

**Comment by jsnmoon** [comment]
> As this is such a significant change, feels like it might be worth holding off until after the Jan code freeze to merge it. What do you think @jsnmoon? (@bluefuton)

Given that we have no other major changes for Jetpack Search in the queue, I feel pretty comfortable shipping this for January :)

Additionally, holding off on a January release will mean we'll be releasing this change simultaneously with our Woo search changes in February.

**Comment by bluefuton** [comment]
Lovely job. 🎉 Works fantastically well on the whole. 

Running into a bit of wonkiness around filtering. If I choose a filter and then do 'Clear filters', the filters are removed from the query string and 'clear filters' vanishes, but the filters aren't updated in the UI:

Screen Shot 2020-12-22 at 16 39 00

(Don't mind the massive category list! That display issue is fixed in a different PR.)

**Comment by jsnmoon** [comment]
> Running into a bit of wonkiness around filtering. If I choose a filter and then do 'Clear filters', the filters are removed from the query string and 'clear filters' vanishes, but the filters aren't updated in the UI (@bluefuton)

I was able to reproduce and fix this issue. Let me know if my latest commit works for you!

**Review by bluefuton — DISMISSED** [review]
> > Running into a bit of wonkiness around filtering. If I choose a filter and then do 'Clear filters', the filters are removed from the query string and 'clear filters' vanishes, but the filters aren't updated in the UI (@bluefuton)
> 
> I was able to reproduce and fix this issue. Let me know if my latest commit works for you!

Filtering now works as expected 👍 Looks good!

**Review by bluefuton — DISMISSED** [review]
✅ Works well for me.

**Comment by bluefuton** [comment]
Sorry @jsnmoon - needs another (small) rebase as the sort PR merged.

**Review by jeherve — APPROVED** [review]
This tests well for me. 🚢

**Comment by jsnmoon** [comment]
r219765-wpcom

**Pull request merged**
